[12.0] Importação de NF-e por XML

lfdivino opened the pull request:
Olá Pessoal,

Este PR visa viabilizar a importação de NF-e:

- Importação do arquivo XML manualmente

Ele é uma separação da importação manual do PR #1355 

@rvalyi @renatonlima @marcelsavegnago @mileo @gabrielcardoso21 Criticas e sugestões são sempre bem vindas

mileo committed:
[FIX] Import NF-e XML

gabrielcardoso21 committed:
[ADD] Wizard to import XML

gabrielcardoso21 committed:
[FIX] Decoding field

gabrielcardoso21 committed:
[FIX] Import

gabrielcardoso21 committed:
[ADD] Menu to import XML

mileo committed:
[FIX] Wizard on manifest

lfdivino committed:
[FIX] Missing fields in NF-e for importing

lfdivino committed:
[FIX] Validation if line fields are not None

lfdivino committed:
[FIX] cEANTrib tag as product's barcode

lfdivino committed:
[ADD] Imported XML to attachments

rvalyi commented:
OK, como eu ja comentei eu acho importante ter esse sistema minimalista. Estrategias de depare ou de importação usando o DFe podem ser plugadas em cima disso em módulos opcionais, mas assim me parece no caminho certo, eu bolei esse dry_run exactamente para isso...

marcelsavegnago commented:
> OK, como eu ja comentei eu acho importante ter esse sistema minimalista. Estrategias de depare ou de importação usando o DFe podem ser plugadas em cima disso em módulos opcionais, mas assim me parece no caminho certo, eu bolei esse dry_run exactamente para isso...

@rvalyi seria melhor a importação ficar em um modulo separado do l10n_br_nfe ? tipo l10n_br_nfe_import ou nao se faz necessário ?

rvalyi commented:
> > OK, como eu ja comentei eu acho importante ter esse sistema minimalista. Estrategias de depare ou de importação usando o DFe podem ser plugadas em cima disso em módulos opcionais, mas assim me parece no caminho certo, eu bolei esse dry_run exactamente para isso...
> 
> @rvalyi seria melhor a importação ficar em um modulo separado do l10n_br_nfe ? tipo l10n_br_nfe_import ou nao se faz necessário ?

Eu acho melhor sim. Mas dá para ter um wizardzinho assim com 150 linhas, isso TB não cria nenhum monstro desse jeito. O foco mesmo deve ser em mapear os impostos na importação.

renatonlima reviewed (changes requested):
O wizard de importação deveria ser herdado do wizard: https://github.com/OCA/l10n-brazil/blob/12.0/l10n_br_fiscal/wizards/base_wizard_mixin.py#L8

A ideia é implementar a importação via arquivo (e futuramente via download) mas utilizando o mesmo wizard.

![Captura de Tela 2021-11-08 às 15 17 39](https://user-images.githubusercontent.com/211005/140802826-ba247ed8-7711-4801-ba72-0c6373403f52.png)

Permitindo as opções:

![Captura de Tela 2021-11-08 às 15 17 52](https://user-images.githubusercontent.com/211005/140802874-9cbae377-b6be-4889-91c2-946bba019c15.png)

A ter uma pré-visutalização:

![Captura de Tela 2021-11-08 às 15 23 11](https://user-images.githubusercontent.com/211005/140803415-7f4cbfad-5a17-439c-9b8c-91d35057b14b.png)

rvalyi reviewed (changes requested):
@lfdivino como eu disse nao vejo grande problema conceitual (so fazer o que o @renatonlima pediu). Mas tem que botar o Travis e Runbot verde tb e ver detalhadamente se nao veio algum lixo junto da branch que vc usou nesse PR.

@renatonlima sobre a pre-visualização tem que ver o que é facil razoavel fazer nesse modulo. Porque talvez isso ja é mais trabalho e tem que ser feito em outros modulos que vaõ herdar isso.

lfdivino committed:
[FIX] Inherit from Fiscal Base Wizard Mixin

lfdivino committed:
[FIX] Active model validation on default_get()

lfdivino committed:
pre-commit

lfdivino committed:
[FIX] Char relational import field insensitive

antoniospneto commented:
@lfdivino tem que dar uma olhada que da forma que está o campo **carrier_id** está em conflito com o **carrier_id** que já existe definido no módulo [l10n_br_delivery](https://github.com/OCA/l10n-brazil/blob/7aeea0cfd3cfad8ccbfc50eb65c149b4612c04aa/l10n_br_delivery/models/fiscal_document_mixin.py#L23) 

Aqui na PR o **carrier_id** está relacionado a um **partner_id** mas no módulo **l10n_br_delivery** está relacionado a um **delivery.carrier**

Por causa disso acaba estourando esse erro:
```
psycopg2.IntegrityError: insert or update on table "l10n_br_fiscal_document" violates foreign key constraint "l10n_br_fiscal_document_carrier_id_fkey"
DETAIL:  Key (carrier_id)=(10) is not present in table "res_partner".
```
Obs: o teste foi feito em um banco de dados que já tem o módulo l10n_br_delivery com dados inseridos.

mileo commented on l10n_br_nfe/wizards/import_document.py:
```diff
@@ -0,0 +1,62 @@
+            "res_id": edoc.id,
+            "res_model": "l10n_br_fiscal.document",
+            "type": "ir.actions.act_window",
+            "target": "new",
```
Não acho legal abrir em uma nova janela, melhor abrir em um formulário.

mileo commented:
Pessoal ta muito estranho esse erro do travis, alguém tem alguma ideia?

gabrielcardoso21 commented:
> Pessoal ta muito estranho esse erro do travis, alguém tem alguma ideia?

Cara, um dos erros que analisei foi um de ordenação dos usuários. No código mesmo não tem algo que possa ter mudado isso, então meu palpite é na alteração feita incluindo o before_script no arquivo .travis.yml. Talvez, ao criar as bases de dados antes do travis e instalar a extensão unaccent a ordenação padrão dos registros do core esteja sendo alterada. Não sei se isso resolveria todos os erros, pois apenas dois dos 7 erros parecem ter alguma relação com essa ordenação dos registros.

mileo committed:
[FIX] Unaccent

mileo committed:
[WIP] Unaccent

mileo commented:
Depends of https://github.com/OCA/maintainer-quality-tools/pull/701

lfdivino committed:
[FIX] Remove unnecessary field relational

felipemotter commented:
Oi, essa PR já está pronta para revisão? 

Estava testando um pouco e vi que ao importar não é possível conferir os dados antes do numero e série da nota serem lançados. 

Na imagem a seguir, eu fiz uma importação mas fechei antes de confirmar. Só que ao tentar importar novamente o mesmo XML, foi acusado esse erro. PS: o documento e a fatura não aparecem criados.

![image](https://user-images.githubusercontent.com/6812128/144935087-c818d609-c788-4455-b399-032e8e47fc14.png)

Outra coisa, ao importar, automaticamente os itens e parceiro foram salvos sem mesmo conferir/confirmar.

mileo commented:
> 

O documento tem que ficar como provisório no módulo l10n_br_fiscal. 

Entretanto devido a arquitetura atual, se vc tiver com o l10n_br_account instalado provavelmente não poderá ver o documento. Em um segundo pull request deve ser feita a integração com o faturamento.

Portanto se for testar testa só com o l10n_br_fiscal instalado e ai posterior a isso podemos propor novas melhorias em PR separados, por exemplo integração com 

1. account;
2. stock;
3. purchase;
4. importação do xml direto do sefaz;

felipemotter commented:
Obrigado pelas explicações @mileo

lfdivino committed:
[ADD] Import Menu in Fiscal Manager group

lfdivino committed:
[FIX] Comodel name for NFe fields

lfdivino committed:
[FIX] Import NFe XML functionalities

lfdivino committed:
[ADD] Previsualization of NFe XML content

mileo commented:
@lfdivino eu acho que o wizard de importação deve ser simples, sem considerar os itens, somente o cabeçalho da nota.

Pois tem que ser simples o suficiente para importar XML de outros tipos de documentos fiscais: CT-E, NFC-E, SAT e etc.

mileo commented:
Inclusive parte do wizard deve ser implementada no fiscal e a parte concreta da NF-e no módulo da NF-e.

hirokibastos committed:
[ADD] Production selection to import wizard.
	- Added a new field to l10n_br_nfe.import_xml.products, where you can choose which product from database that new corresponds to.
	- Added new nfe40_xProd field to document line so that product import works correctly
		- Product importing now matches product chosen on wizard's new field.
		- No longer creates a new product every time an import in made.
	- Changed the value of some fields in xml file when importing, so they match internal values.
	- Added validation for ncm, nbm and cest values when importing.

hirokibastos committed:
[FIX] Changed l10n_br_delivery_nfe field name so import bug stops happening.
	- Using the partner_id field during importing caused a problem because document.py has a field with the same name.

hirokibastos committed:
[FIX] Add missing file for last commit

hirokibastos committed:
[ADD] Same key validation and NCM, NBM and Cest validation for each product.
	- Added validation to check if a document with the same key as the nfE being imported is alredy in the database
		- Raise error if key is found in database (same NFe cant be imported more than once)
	- Added validation to check if the values for NCM, NBM and Cest in the xml file and for the products selected for each line match.
		- Raise error if they don't

hirokibastos committed:
[ADD] Create relation between product.product, suplier and suplier's product name.
	- Creates/updates a product.supplierinfo relation when importing a NFe
	- Searches for this relation and automatically fills imported_products_ids.product_id when choosing a xml file.

hirokibastos committed:
[RMV] Removed 'Confirm' button for imported documents.

hirokibastos committed:
[ADD] Added new 'Confirm' button for imported documents.
	- Added usual 'Confirm' button shown when document state is 'em_digitacao' for imported documents
	- Added new 'Confirm' button for imported documents
		- This button sets the state of the document to 'autorizada'
	- Hid 'Cancel', 'Correction' and 'Return' buttons shown when document state is 'autorizada' for imported documents

hirokibastos committed:
[FIX] Fixed info fields not being imported.

hirokibastos committed:
[ADD] Sets partner as a supplier when importing a NFe

hirokibastos committed:
[FIX] Error when trying to create a res.partner and XML file has empty xFant field.

hirokibastos committed:
[UPD] Rollback changes that forced imported document state to be 'authorized'

hirokibastos committed:
[FIX] Fixed UOM never finding a match

hirokibastos committed:
[FIX] Fix error when importing document with <entrega> field
	- Now creates a delivery adress contact with the information within the field.

hirokibastos committed:
[UPD] Check if fields exist before attributions for <entrega> field
	- Check if each field is in rec_dict before trying to insert it's value in vals dict

hirokibastos committed:
[RMV] 'Confirm Import' button and some commented code

hirokibastos committed:
[ADD] Added comments to new match_or_create_m2o method for res.partner

hirokibastos committed:
[FIX] Recursive relation

hirokibastos committed:
[UPD] Moved transporter import fix from l10n_br_delivery_nfe to l10n_br_nfe

hirokibastos committed:
[ADD] Internal UOM field to import wizard
	- Add a UOM field to import wizard
	- The internal UOM equivalent to the XML UOM should be selected in this field
	- Selected UOM is saved in product.supplierinfo after import is done

hirokibastos committed:
[FIX] Bypass or maybe fix error when validation invoice for imported document
	- When importing a document from a partner from 'Simples Nacional', choise11 field is left empty

hirokibastos committed:
[UPD] Delivery Adress no longer updated by importing document
	- A new delivery adress will be created when a match is not found.

hirokibastos committed:
[UPD] Created separeted method for editting parsed_xml

hirokibastos committed:
[RMV] Removed wrong fix for icms_tag bug

hirokibastos committed:
[FIX] Passing instance instead of id in create method

hirokibastos committed:
[FIX] Mistake in last commit

hirokibastos committed:
[FIX] Change search parameter for supplier info from product_name to product_id
	- Changed how supplierinfo record update is done when match is found
	- Also removed readonly from product_uom field for product.supplierinfo

hirokibastos committed:
[FIX] Now imports CSOSN when xml contains ICMSSN fields

hirokibastos committed:
[ADD] Better conditions to match 'entrega' field in XML file
	- Also hardcoded Brazil's code when searching for a UF
		- This was done because XML may be missing cPais field

hirokibastos committed:
[FIX] Now considers UoM when saving product.supplierinfo price

hirokibastos committed:
[RMV] Remove trash left when fixing transporter bug and run pre-commit

hirokibastos committed:
[FIX] Add product_id to supplierinfo when creating it from purchase order.

hirokibastos committed:
[FIX] Remove last fix and add one that actually works

hirokibastos committed:
[ADD] Now doesn't try to create country if none is found

hirokibastos committed:
[UPD] Improved how creation of delivery adress is made

hirokibastos committed:
[FIX] Change how default value for uom in supplierinfo is obtained

rvalyi commented:
@lfdivino seria interessante fazer uns `git  squash`nessa branch, porque o tamanho do PR é ate razoável. Mas 62 commits com muitos que é 3 passos para frente seguido de 2 passos para traz vai ficar quase impossível fazer o cherry-pick para mandar na 14.0 depois. Não seria possível quebrar isso em uns 10-15 commits?

marcelsavegnago commented on l10n_br_nfe/wizards/import_document.py:
```diff
@@ -0,0 +1,333 @@
+    uom_internal = fields.Many2one(
+        "uom.uom",
+        "Internal UOM",
+        help="Unidade de medida interna equivalente à unidade de medida no documento"
```
Não sei se todos concordam, mas não seria melhor deixar as strings e helps em inglês? Acredito que boa parte do código de outros módulos já foi feito um tratamento e padronizado para o inglês.

cc @rvalyi @renatonlima @netosjb @felipemotter

marcelsavegnago replied:
Obs. Entendo que hoje temos um problema com o weblate da v12 da localização.

hirokibastos replied:
Resolvido

rvalyi commented on .travis.yml:
```diff
@@ -49,7 +57,7 @@ env:
 
 install:
-  - git clone --depth=1 https://github.com/OCA/maintainer-quality-tools.git
+  - git clone --depth=1 https://github.com/kmee/maintainer-quality-tools.git --branch=feature/setup-template
```
isso é por causa do unaccent? Vc não consegue adaptar os testes para rodar minimalmente mesmo sem o unaccent? Ainda que seja com um mock da vida?

Para ser sinceiro, eu tenho um problema com esse fork para a manutenção depois. A experiencia mostra que a KMEE ta muito enrolada para manter coisas que ela faz em repos externos (erpbrasil.assinatura, erpbrasil.edoc, libs de nfse, migração dos módulos da KMEE para a v14...) eu vejo um grande risco aqui de começar a depender de um repo de teste sem manutenção e depois criar ainda mais problemas na hora da migração para a v14...

rvalyi replied:
eu comentei aqui sobre esse PR no maintainer-quality-tools: https://github.com/OCA/maintainer-quality-tools/pull/701#issuecomment-1081268541

marcelsavegnago replied:
@lfdivino Como ficou esta questão ?

mileo replied:
Vamos ter que implementar no CI novo, vou ver como implementar assim que possível.

mileo replied:
Verifiquei que o módulo l10n_br_nfe ainda não foi migrado para 13 e 14, imagino que isso ainda vai levar alguns meses. Será que podemos avançar com esse PR e também com o https://github.com/OCA/maintainer-quality-tools/pull/701#issuecomment-1081268541

E também estamos usando o travis na 14. E pelo que analisei no código do runboat ele é mais simples e menos parametrizável que o maintainer-quality-tools mantendo o mesmo db, o que aparentemente não causará mesmo problema, mas é preciso testar pra ter certeza.

DiegoParadeda committed:
[REF] cleaner import code

marcelsavegnago commented:
@hirokibastos consegue aumentar a cobertura de testes ? o que acha ?

marcelsavegnago commented:
@hirokibastos @lfdivino consegue rodar o pre-commit ? o interessante eh rodar sempre o pre-commit antes de comitar

marcelsavegnago commented:
@hirokibastos vc consegue dar uma olhada neste link e fazer o processo indicado por ele ? caso já tenha feito só da um toque aqui fazendo favor
https://odoo-community.org/page/cla

rvalyi commented:
@lfdivino @hirokibastos @mileo eu tambem insisto na questão que fazer uns git squash e de quebrar essa PR em pedaços menores, tem coisas que são faceis de aprovar, basta extrair algumas mudanças simples primeiro. Senão vai ser impossivel fazer os cherry-picks pro port depois, eu ja tinha comentado sobre isso um mes atras https://github.com/OCA/l10n-brazil/pull/1701#issuecomment-1081075202

Vcs não podem esperar que a gente faça todo trabalho de arquitetura, de manutenção de migração no projeto e depois exigir que a gente faça merge de PRs de mais 1000 linhas ou 70 commits que não passou por nenhum crivo de qualidade e depois largar isso no repo e passar a outra coisa, como já foi feito tantas vezes... cc @marcelsavegnago @netosjb @renatonlima

hirokibastos committed:
[ADD] Added tests and made some minor fixes

marcelsavegnago commented:
> @hirokibastos vc consegue dar uma olhada neste link e fazer o processo indicado por ele ? caso já tenha feito só da um toque aqui fazendo favor https://odoo-community.org/page/cla

@hirokibastos você chegou a dar uma olhada neste link ?

hirokibastos commented:
> > @hirokibastos vc consegue dar uma olhada neste link e fazer o processo indicado por ele ? caso já tenha feito só da um toque aqui fazendo favor https://odoo-community.org/page/cla
> 
> @hirokibastos você chegou a dar uma olhada neste link ?

@marcelsavegnago Está feito, perdão pela demora com a resposta.

lluisgustavoreis commented:
> @lfdivino @hirokibastos @mileo eu tambem insisto na questão que fazer uns git squash e de quebrar essa PR em pedaços menores, tem coisas que são faceis de aprovar, basta extrair algumas mudanças simples primeiro. Senão vai ser impossivel fazer os cherry-picks pro port depois, eu ja tinha comentado sobre isso um mes atras [#1701 (comment)](https://github.com/OCA/l10n-brazil/pull/1701#issuecomment-1081075202)
> 
> Vcs não podem esperar que a gente faça todo trabalho de arquitetura, de manutenção de migração no projeto e depois exigir que a gente faça merge de PRs de mais 1000 linhas ou 70 commits que não passou por nenhum crivo de qualidade e depois largar isso no repo e passar a outra coisa, como já foi feito tantas vezes... cc @marcelsavegnago @netosjb @renatonlima

Como você sugere que o PR seja dividido? Pois não vemos impacto da forma como está

antoniospneto commented:
@lluisgustavoreis 

Um exemplo que eu vejo que já poderia ser extraído para um PR separada é a questão do transporte:
![image](https://user-images.githubusercontent.com/634278/170057791-e6bf295a-35f8-46ac-a8aa-93cba68b99d6.png)
![image](https://user-images.githubusercontent.com/634278/170058006-f9e241c4-9957-417e-9b69-78041f2ef8cc.png)
![image](https://user-images.githubusercontent.com/634278/170058116-f9c0b6b3-9eda-4a9d-8761-68ea380dde3b.png)

rvalyi commented:
> @lluisgustavoreis 
> 
> Um exemplo que eu vejo que já poderia ser extraído para um PR separada é a questão do transporte:
> ![image](https://user-images.githubusercontent.com/634278/170057791-e6bf295a-35f8-46ac-a8aa-93cba68b99d6.png)
> ![image](https://user-images.githubusercontent.com/634278/170058006-f9e241c4-9957-417e-9b69-78041f2ef8cc.png)
> ![image](https://user-images.githubusercontent.com/634278/170058116-f9c0b6b3-9eda-4a9d-8761-68ea380dde3b.png)
> 
> 

É um bom exemplo sim @netosjb

hirokibastos committed:
[FIX] Fixed nfe40_transporta relations, so it's imported correctly

marcelsavegnago commented:
@hirokibastos @lfdivino vocês acreditam que cabe alguns squashs para deixar o histórico mais limpo? Olhando por cima me parece que tem commtis corrigindo outros commits da mesma PR.

rvalyi commented:
Vou fechar essa PR porque foi quebrado em PRs menores que já tiveram merge para a grandes maioria deles. Os pequenos que ficaram estão ainda visível para receber revisões e ajustes. Obrigado a todos que participaram!
